Self Service LTK Users

amagana3 opened the pull request:
This PR adds the ability to do some self service LTK through Terraform. This has been a long awaited ask from the ALKS issue tracker.

- Ability to add, update, and delete LTK user.
- Added example in the examples directory.
- Added some tests.

webbbarker commented on resource_alks_ltk.go:
```diff
@@ -0,0 +1,98 @@
+func populateResourceDataFromLTK(longTermKey *alks.GetLongTermKeyResponse, d *schema.ResourceData) error {
+	d.SetId(longTermKey.UserName)
+	_ = d.Set("access_key", longTermKey.AccessKeyID)
+	_ = d.Set("create_date", longTermKey.CreateDate)
```
This isn't present in your Schema definition or in the state output of `create`.  You can avoid that by following the recommended pattern of the `create` method calling `return readTheResourceMethod()`, but you also need to consider the state that's not available when you do that, like the `secret_key`.

amagana3 commented:
### Update
Okay so I've done a couple of things to this resource that I didn't realize I had left off. I've added an `exists` functionality, which simply checks if the resource exists by the ID (the LTK name). I've also added an `import` functionality which will be of big use especially for those who already have LTKs in their accounts which were not tracked before.

The second thing, I am not sure how to handle the `secret_key` and `access_key` in the state files. I've looked into some of the docs and found this: https://www.terraform.io/docs/extend/best-practices/sensitive-state.html#using-the-sensitive-flag. I will say this though, if you have an LTK which has been imported using `import`, the secret and access keys are actually not stored in the state interestingly enough.

@webbbarker - Some notes for when you're testing locally ^

webbbarker commented on resource_alks_ltk.go:
```diff
@@ -11,6 +11,11 @@ func resourceAlksLtk() *schema.Resource {
 		Read:   resourceAlksLtkRead,
 		Delete: resourceAlksLtkDelete,
+		Exists: resourceAlksLtkExists,
+		Importer: &schema.ResourceImporter{
```
FWIW, I believe this is only used if you are changing the state format version.  We used something similar when we added machine identities as we had a new state field that we needed to set for existing roles.

amagana3 replied:
Ah I see now after reading some more on it. Do you think it is worth keeping this `exists` implementation, or maybe removing it to prevent confusion? I cannot imagine we'd change the state field any time soon...

webbbarker replied:
Let's leave it alone for the time being.  No harm there.

amagana3 commented:
### Update
When adding the `import` it uses the `read` function to do its checks. Essentially it looks to see if the `id` is consistent throughout the state, but since we were previously _setting_ the ID in our read function, there was a mismatch and therefore re-creating the LTK when doing a `plan` after the `import`. This has been fixed by setting the state's field for `iam_username` rather than re-setting the `id` which _only_ the `create` function should do.

webbbarker commented on resource_alks_ltk.go:
```diff
@@ -78,7 +78,9 @@ func resourceAlksLtkRead(d *schema.ResourceData, meta interface{}) error {
 	log.Printf("[INFO] alks_ltk.id: %v", d.Id())
 
-	return populateResourceDataFromLTK(resp, d)
+	_ = d.Set("iam_username", resp.UserName)
+
```
I think you should be setting as much information here as possible so that the `read` operation fulfills its contract.  So basically, add in the `access_key`.  In the future we should consider returning additional information (like the ARN) back to the user when the ALKS LTK endpoint is called.
